src/node_snapshotable.cc

```diff
@@ -1053,8 +1052,6 @@ static void ResetContextSettingsBeforeSnapshot(Local<Context> context) {
   context->AllowCodeGenerationFromStrings(true);
 }
 
-Mutex SnapshotBuilder::snapshot_data_mutex_;
```

Comment:
We actually have a const_cast at L1064, though I *think* that V8 doesn't actually need `params->snapshot_blob` to be non-const. Can you modify the `v8::CreateParams::snapshot_blob` to a const pointer too and see if it works? If so we should send a patch to the upstream first to get rid of the const cast, so that they don't start to need it to be non-const without us knowing about it.

Reply:
Uh-oh, it appears v8 is const-casting too.. 😅  https://github.com/nodejs/node/blob/50930a0fa08297d0ce7e67fa6594fe47937b99ff/deps/v8/src/snapshot/snapshot-data.h#L77

Reply:
Actually, V8’s `StartupData` is just a `const char* data` + `int raw_size`. That can’t really be mutated by V8 anyway.

Reply:
Yeah, it looks like V8 isn't actually mutating the data, just reusing the data structure that is also used for serialization. Perhaps some additional field in `SerializedData` that can be used to DCHECK mutation would be enough.

Reply:
(And yes our `const_cast` was only there because the V8 API wants a non-const pointer)

Reply:
>  If V8 were to mutate the blob, we would need to learn about that; and an easy way to do that is to make the actual data const so that attempting to mutate it crashes.

Isn't mutating const-casted data undefined behavior, so it's not certain that it would crash (or it would crash reliably)? And that's my main concern because we might be seeing crashes that can't be easily link to the mutation, depending on how the mutation happens..it'd probably be safer to have some DCHECK against the mutation, either in V8 or we DCHECK a verification of the snapshot checksum when we initialize a new isolate with the snapshot data.

Reply:
> V8 really can’t mutate the blob, since it needs to be re-usable across Isolates (both concurrently and sequentially).

I think you are talking about the default snapshot, not the one passed into `IsolateParams`? V8 also has a mutex to guard against its default snapshot read from disk: https://github.com/nodejs/node/blob/6bd756d7c6dfb7dc25daee329ad70df68c14223e/deps/v8/src/snapshot/snapshot-external.cc#L57 (not entirely sure if that's strictly necessary for V8 either)

Reply:
(On a second thought, why is the mutex used in an accessor (both in our code and in V8)? It probably should've been somewhere surrounding `Isolate::Initialize()` instead to serve that purpose..

Reply:
Oh I see, the V8 mutex is guarding against refreshing the snapshot while it's being read for isolate initialization, which isn't a thing for us because we don't refresh our embedded snapshot blob in any way. The one we read from disk is in a separate location. So we really don't need this mutex anyway (and if we really want to avoid a race from mutation from V8, we should have a mutex surrounding `Isolate::Initialize()` instead).

Reply:
> Isn't mutating const-casted data undefined behavior, so it's not certain that it would crash (or it would crash reliably)?

From a language perspective, yeah, it's UB. I was honestly expecting that marking this as `const` would put it in the .rodata section of the binary; looking at the compiled result, it seems like it's a bit too complex for that.

In any case, the V8 CL is merged now, if you'd like me to cherry-pick the `IsolateParams` change and remove the `const_cast` I'm happy to also do that.

> DCHECK a verification of the snapshot checksum

Just to avoid disambiguity, this PR does not affect the const-ness of the actual snapshot data; that's already a `const char*` anyway (which also does actually end up in .rodata). If V8 does `const_cast` on that, then yeah, as you point out, it's on them to make sure they do it properly.